---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# qPCR.CFX.process

<!-- badges: start -->

<!-- badges: end -->

The goal of the R package `qPCR.CFX.process` is to facilitate the processing of real-time quantitative PCR results, exported by the Bio-Rad CFX Connect software. It is meant for relative quantification of target genes, and supports one or multiple reference genes.

Its input is calculated quantification cycle (Cq) values and its output is relative quantity to a designated control sample. If standard curve reactions are included, primer efficiencies are calculated and the method established by [Pfaffl (2001)](#ref1) is applied to calculate relative quantity, using the *common base* calculation approach, described by [Ganger et al. (2017)](#ref2). Otherwise, the [Livak and Schmittgen (2001)](#ref3) method is applied, assuming all primer efficiencies to be equal to 100%.

## Installation

You can install the `qPCR.CFX.process` package from [GitHub](https://github.com/) by using the `devtools` package. If `devtools` is not installed in your system, install it with the following line:

```{r,eval=FALSE}
install.packages("devtools")
```

When `devtools` is installed, run the following line to install the `qPCR.CFX.process` package:

```{r,eval=FALSE}
devtools::install_github("dimitriskokoretsis/qPCR.CFX.process")
```

## Setting up the real-time qPCR run

The `qPCR.CFX.process` package makes certain assumptions about the structure of the data. Naming conventions in your setup should comply with these assumptions. The following rules should be followed:

-   Primer efficiencies are calculated from standard curves, based on serial dilutions of the template. The starting quantity input by the user is in arbitrary units, but the dilution rate is important. For example, input starting quantities of 5, 1, 0.2 and 0.04 should give the same efficiency results as 10, 2, 0.4 and 0.08 (both describe a dilution rate of 5).

-   The "sample" field is used both for the name of the reaction template and for its biological replicate number. The biological replicate number should come last and be separated with a space from the sample's name. For example: "template name X", where X is the biological replicate number. Sample name doesn't matter for calculations in standard curve reactions, as the same template is always used. Use a descriptive name for your convenience.

-   Technical replicates should have the exact same names in all fields. They are essentially copies of the same reaction and their Cq values will be averaged.

## Example

We will demonstrate the functionality of `qPCR.CFX.process` with an example real-time qPCR result. In this mock experiment, the expression of **2 genes of interest** is measured on a **test condition**, in relation to a **control condition**. There are **3 biological replicates** for each condition and **one reference gene** is used.

### Import real-time qPCR results to the R environment

The Bio-Rad CFX software exports a number of different comma-separated files (.csv extension) per run. `qPCR.CFX.process` handles the file whose name ends in "Quantification Cq Results_0.csv". This file contains information on type of reaction, sample, amplification target, and calculated quantification cycle (Cq) per well.

There are different ways to import a CSV file into the R environment. We will use the `fread` function of the `data.table` package to import the data in a `data.table`, which we will call `Cq.data`.

```{r}
# Import data from file
Cq.data <- data.table::fread("test_data/username_date_time_model - Quantification Cq Results_0.csv",check.names=TRUE)

# Display data
knitr::kable(Cq.data)
```

The `check.names=TRUE` argument of `fread` converts the spaces to dots in the field names, to minimize downstream processing errors. Each row (record) of the table corresponds to a single well, and each column (field) represents the following information:

-   `Target`: The gene being amplified - essentially, the primer pair used in the reaction.

-   `Content`: The type of reaction. It is set during setup with the Bio-Rad CFX Connect software and can have one of 3 values:

    -   "NTC": non-template control

    -   "Std": standard curve reaction

    -   "Unkn": unknown sample reaction

-   `Sample`: The reaction template and its biological replicate number.

-   `Cq`: The calculated quantification cycle of the reaction.

-   `Starting.Quantity..SQ`: Only for standard curve reactions, the template's starting quantity as was input during reaction setup.

-   `Log.Starting.Quantity`: Only for standard curve reactions, the base-10 logarithm of the starting quantity.

Actual exported files from Bio-Rad CFX connector contain more fields, but this is all the information needed for the expression analysis.

### Process real-time qPCR results

To process the imported results, load the `qPCR.CFX.process` package and feed the primary data to the `qPCR_analysis_wrap` function. This will perform the whole processing in one go. Set the arguments `refgene` to the name of your reference gene and `control` to the name of your control condition.

```{r}
# Load the qPCR.CFX.process package
library(qPCR.CFX.process)

# Process qPCR results and store them in the qPCR.analysis variable as a list
qPCR.analysis <- Cq.data |>
  qPCR_analysis_wrap(refgene="reference.gene",
                     control="control.condition")
```

The qPCR results have now been processed and are contained in a list named `qPCR.analysis`.

### Investigate real-time qPCR results

To check the contents of `qPCR.analysis`, you can use the `str` function (as in "structure"). The `max.level` argument is set to 1, to only show the first level of the list's structure.

```{r}
str(qPCR.analysis,max.level=1)
```

Therefore, the `qPCR.analysis` list has 4 elements:

-   `NTC`: A `data.table` with primer targets and Cq values, each row being a non-template control reaction.

-   `std.curve`: A list of 3 elements. See [Standard curve calculations] section for more details.

-   `unk.rxn`: A `data.table` with Cq values for each sample, biological replicate and gene.

-   `expression`: A `data.table` with the calculated expression of each sample, biological replicate and target gene.

#### Non-template controls

To see the `NTC` `data.table` of the `qPCR.analysis` list, call it as follows:

```{r}
# Display NTC data.table
knitr::kable(qPCR.analysis$NTC)
```

As we see, the non-template control reaction for gene.of.interest.1 has a Cq value of 34.9817 - while the other two non-template control reactions did not yield a Cq value (`NaN` - not a number).

Ideally, all three should be `NaN`. However, as the Cq value is quite high, we can move forward.

#### Standard curve calculations

The `std.curve` element of the `qPCR.analysis` list contains the results of standard curve reactions and primer efficiencies and is itself a list of 3 elements. We can check its structure as follows:

```{r}
str(qPCR.analysis$std.curve,max.level=1)
```

-   `data`: A `data.table` with the Cq values against log(Starting quantity) for each target gene.

-   `efficiencies`: A `data.table` with the calculated efficiencies for each target gene.

-   `plot`: A `ggplot2`-based scatter plot with drawn standard curves.

```{r}
# Display "data" data.table
knitr::kable(qPCR.analysis$std.curve$data)
```

In the `data` `data.table`, each row corresponds to a target gene and starting quantity of template. The `Cq.average` and `Cq.st.dev` fields are the arithmetic mean and standard deviation of technical replicates, respectively.

```{r}
# Display "efficiencies" data.table
knitr::kable(qPCR.analysis$std.curve$efficiencies)
```

In the `efficiencies` `data.table`, each row corresponds to a target gene. The `slope` field is the slope of the regression line when Cq is plotted against log(starting quantity). The `efficiency` field shows the calculated primer efficiency (1 corresponds to 100%) and `amplification.base` is the calculated rate of increase per cycle (efficiency + 1).

```{r}
# Display plot
qPCR.analysis$std.curve$plot
```

The `plot` element is a scatter plot with quantitation cycle against log(starting quantity), with regression lines for each target gene. Data points are the arithmetic means between technical replicates, while error bars show their standard deviation.

#### Unknown reactions

The `unk.rxn` `data.table` gathers and summarizes the unknown sample reactions for each target gene, sample, and biological replicate.

```{r}
# Display "unk.rxn" data.table
knitr::kable(qPCR.analysis$unk.rxn)
```

-   `Sample`: Name of sample template

-   `Biol.rep`: Biological replicate number

-   `Target`: Name of target gene

-   `Cq.tech.mean`: Arithmetic mean between technical replicates

-   `Cq.tech.sd`: Standard deviation between technical replicates

#### Calculated expression

The `expression` `data.table` is the final result of the real-time qPCR data processing. Each row corresponds to a sample, biological replicate and gene of interest.

```{r}
# Display "expression" data.table
knitr::kable(qPCR.analysis$expression)
```

There are many fields in this table to depict calculation stages, but the most important ones are the following:

-   `Sample`: Name of sample template

-   `Target`: Name of gene of interest

-   `fold.change`: Fold-change of quantity in relation to control sample

### Visualize expression results

We can visualize the fold-change between the control and test conditions for each gene of interest.

Any visualization tool may be used. The `bar_point_plot` function of the [`datavis`](https://github.com/dimitriskokoretsis/datavis) package is quite easy to use for this task. To install it in your system, follow the instructions in its page.

Importantly, for fold-change of quantity, mean and variation parameters only make sense in the logarithmic scale. By setting the `mean.type` argument to "geometric", the mean and standard deviation are calculated on log-transformed fold-change and then exponentiated back to their original scale. These are the *geometric* mean and standard deviation.

```{r}
# Load datavis package
library(datavis)

# Create bar plot
expression.plot <- qPCR.analysis$expression |>
  bar_point_plot(x="Target", # X axis grouping based on "Target" field
                 y="fold.change", # Y axis value is "fold.change" field
                 color.group="Sample", # Color grouping based on "Sample" field
                 mean.type="geometric") # Calculate geometric mean and SD, instead of arithmetic

# Show bar plot
expression.plot
```

## References

<a name="ref1">Pfaffl, M.W. (2001). A new mathematical model for relative quantification in real-time RT–PCR. Nucleic Acids Research, 29 (9), e45. <https://doi.org/10.1093/nar/29.9.e45></a>

<a name="ref2">Ganger, M.T., Dietz, G.D. & Ewing, S.J. (2017). A common base method for analysis of qPCR data and the application of simple blocking in qPCR experiments. BMC Bioinformatics, 18. <https://doi.org/10.1186/s12859-017-1949-5></a>

<a name="ref3">Livak, K.J. & Schmittgen, T.D. (2001). Analysis of Relative Gene Expression Data Using Real-Time Quantitative PCR and the 2−ΔΔCT Method. Methods, 25 (4), 402–408. <https://doi.org/10.1006/meth.2001.1262></a>
